tests/phpunit/tests/kses.php

```diff
@@ -1362,12 +1362,73 @@ public function data_safecss_filter_attr() {
-		$expected = '<div data-foo="foo" data-bar="bar" data-two-hyphens="remains">Pens and pencils</div>';
+		$test     = '<div data-foo="foo" data-bar="bar" datainvalid="gone" data--double-dash="retained"  data-trailing-dash-="allowable" data-two-hyphens="remains">Pens and pencils</div>';
+		$expected = '<div data-foo="foo" data-bar="bar" data--double-dash="retained" data-trailing-dash-="allowable" data-two-hyphens="remains">Pens and pencils</div>';
 
 		$this->assertSame( $expected, wp_kses_post( $test ) );
 	}
 
+	/**
+	 * Ensures proper recognition of a data attribute and how to transform its
+	 * name into what JavaScript code would read from an element's `dataset`.
+	 *
+	 * @ticket 61052
+	 *
+	 * @dataProvider data_possible_custom_data_attributes_and_transformed_names
+	 *
+	 * @param string      $attribute_name      Raw attribute name.
+	 * @param string|null $dataset_name_if_any Transformed attribute name, or `null`
+	 *                                         if not a custom data attribute.
+	 */
+	public function test_wp_kses_transform_custom_data_attribute_name_recognizes_data_attributes( $attribute_name, $dataset_name_if_any ) {
+		$transformed_name = wp_kses_transform_custom_data_attribute_name( $attribute_name );
+
+		if ( isset( $dataset_name_if_any ) ) {
+			$this->assertNotNull(
+				$transformed_name,
+				"Failed to recognize '{$attribute_name}' as a custom data attribute."
+			);
+
+			$this->assertSame(
+				$dataset_name_if_any,
+				$transformed_name,
+				'Improperly transformed custom data attribute name.'
+			);
+		} else {
+			$this->assertNull(
+				$transformed_name,
+				"Should not have identified '{$attribute_name}' as a custom data attribute."
+			);
+		}
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[].
+	 */
+	public static function data_possible_custom_data_attributes_and_transformed_names() {
+		return array(
+			// Non-custom-data attributes.
+			'Normal attribute'             => array( 'post-id', null ),
+			'Single word'                  => array( 'id', null ),
+
+			// Normative custom data attributes.
+			'Normal custom data attribute' => array( 'data-post-id', 'postId' ),
+			'Leading dash'                 => array( 'data--before', 'Before' ),
+			'Trailing dash'                => array( 'data-after-', 'after-' ),
+			'Double-dashes'                => array( 'data-wp-bind--enabled', 'wpBind-Enabled' ),
+			'Triple-dashes'                => array( 'data---one---two---', '-One--Two---' ),
+
+			// Unexpected but recognized custom data attributes.
+			'Only comprising a prefix'     => array( 'data-', '' ),
+			'With upper case ASCII'        => array( 'data-Post-ID', 'postId' ),
+			'With Unicode whitespace'      => array( "data-\u{2003}", "\u{2003}" ),
+			'With Emoji'                   => array( 'data-🐄-pasture', '🐄Pasture' ),
+			'Brackets and colon'           => array( 'data-[wish:granted]', '[wish:granted]' ),
```

Comment:
WordPress can be stricter than the specs and browser implementations (especially that latter if they differ from the former), to an extent that's the entire purpose of KSES.

I don't think accounting for special characters is overly wise as it's too easy to miss situations in which a contributor or author role can break out of the intended behaviour.

Reply:
thanks @peterwilsoncc! in the code for this PR there are two stages: the first is determining if there's a data attribute, and then the second is locking down a smaller allowable subset. these tests, and this function, is to ensure that we can all agree on what _is_ a data attribute before we then apply further constraints.

it doesn't have to be this way, of course, but I find that it's really easy to let things slide through when we conflate the two ideas. in this particular case it looks like everything else should be eliminated in `wp_kses_attr_check()` other than the specific data attributes, but I also had to double and triple check the logic to make sure we weren't letting through some data attributes only because they didn't fit the pattern we expected.

so this was done for the reason of making it less likely to let something slip through. it doesn't have to be done this way, but it was the purpose for the structuring I made.

Reply:
> so this was done for the reason of making it less likely to let something slip through. it doesn't have to be done this way, but it was the purpose for the structuring I made.

Thanks @dmsnell that makes a lot of sense. 👯